From 992869c0726367365d71e9519a0348fd6f8de311 Mon Sep 17 00:00:00 2001
From: Manoj Sekar <quic_sekar@quicinc.com>
Date: Mon, 26 Feb 2024 18:18:21 +0530
Subject: [PATCH] WLANRTSYS-62198 Multi-AP: Generation of Multi-AP Profile
 subelement

Add support for including the Multi-AP Profile subelement into the
Multi-AP element.

Signed-off-by: Manoj Sekar <quic_sekar@quicinc.com>
---
 src/common/ieee802_11_common.c | 13 ++++++++++++-
 src/common/ieee802_11_common.h |  1 +
 src/common/ieee802_11_defs.h   |  6 ++++++
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/common/ieee802_11_common.c b/src/common/ieee802_11_common.c
index f395f4602..e52b2d47b 100644
--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -2744,7 +2744,7 @@ size_t add_multi_ap_ie(u8 *buf, size_t len,
 	u8 *pos = buf;
 	u8 *len_ptr;
 
-	if (len < 9)
+	if (len < 6)
 		return 0;
 
 	*pos++ = WLAN_EID_VENDOR_SPECIFIC;
@@ -2755,10 +2755,21 @@ size_t add_multi_ap_ie(u8 *buf, size_t len,
 	*pos++ = MULTI_AP_OUI_TYPE;
 
 	/* Multi-AP Extension subelement */
+	if (buf + len - pos < 3)
+		return 0;
 	*pos++ = MULTI_AP_SUB_ELEM_TYPE;
 	*pos++ = 1; /* len */
 	*pos++ = multi_ap->capability;
 
+	/* Add Multi-AP Profile subelement only for R2 or newer configuration */
+	if (multi_ap->profile >= MULTI_AP_PROFILE_2) {
+		if (buf + len - pos < 3)
+			return 0;
+		*pos++ = MULTI_AP_PROFILE_SUB_ELEM_TYPE;
+		*pos++ = 1;
+		*pos++ = multi_ap->profile;
+	}
+
 	*len_ptr = pos - len_ptr - 1;
 
 	return pos - buf;
diff --git a/src/common/ieee802_11_common.h b/src/common/ieee802_11_common.h
index 7060e23e9..199179a74 100644
--- a/src/common/ieee802_11_common.h
+++ b/src/common/ieee802_11_common.h
@@ -32,6 +32,7 @@ struct mb_ies_info {
 
 struct multi_ap_params {
 	u8 capability;
+	u8 profile;
 };
 
 /* Parsed Information Elements */
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 743f1f91a..00278b7c8 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -1495,11 +1495,17 @@ struct auth_fail_list {
 #define WFA_CAPA_OUI_TYPE 0x23
 
 #define MULTI_AP_SUB_ELEM_TYPE 0x06
+#define MULTI_AP_PROFILE_SUB_ELEM_TYPE 0x07
+
 #define MULTI_AP_TEAR_DOWN BIT(4)
 #define MULTI_AP_FRONTHAUL_BSS BIT(5)
 #define MULTI_AP_BACKHAUL_BSS BIT(6)
 #define MULTI_AP_BACKHAUL_STA BIT(7)
 
+#define MULTI_AP_PROFILE_1	1
+#define MULTI_AP_PROFILE_2	2
+#define MULTI_AP_PROFILE_MAX	6
+
 #define WMM_OUI_TYPE 2
 #define WMM_OUI_SUBTYPE_INFORMATION_ELEMENT 0
 #define WMM_OUI_SUBTYPE_PARAMETER_ELEMENT 1
-- 
2.43.0

